syntax = "proto3";

option go_package = "jql/jqlpb";

package jql;

service JQL {
	rpc ListRows (ListRowsRequest) returns (ListRowsResponse);
	rpc GetRow (GetRowRequest) returns (GetRowResponse);
  rpc WriteRow (WriteRowRequest) returns (WriteRowResponse);
	rpc DeleteRow (DeleteRowRequest) returns (DeleteRowResponse);
	rpc Persist (PersistRequest) returns (PersistResponse);
}

message EqualMatch {
	string value = 1;
}

message LessThanMatch {
	string value = 1;
}

message GreaterThanMatch {
	string value = 1;
}

message InMatch {
	repeated string values = 1;
}

message ContainsMatch {
	bool exact = 1;
	string value = 2;
}
	

message Filter {
	bool negated = 1;
	string column = 2;
	oneof match {
		EqualMatch equal_match = 3;
		LessThanMatch less_than_match = 4;
		GreaterThanMatch greather_than_match = 5;
		InMatch in_match = 6;
		ContainsMatch contains_match = 7;
	}
}

message Condition {
	repeated Filter requires = 1;
}

message ListRowsRequest {
	string table = 1;
	// Use DNF for maximum expressibility. Filtering is an or clause
	// of and clauses made up of primitive filters.
	repeated Condition conditions = 2;
	string order_by = 3;
	bool dec = 4;
	uint32 offset = 5;
	uint32 limit = 6;
}

enum EntryType {
	STRING = 0;
	INT = 1;
	DATE = 2;
	ENUM = 3;
	ID = 4;
	TIME = 5;
	MONEYAMT = 6;
	FOREIGN = 7;
	FOREIGNS = 8;
}

message Column {
	string name = 1;
	EntryType type = 2;
	int32 max_length = 3;
}

message Entry {
	string formatted = 1;
}

message Row {
	repeated Entry entries = 1;
}

message ListRowsResponse {
	repeated Column columns = 1;
	repeated Row rows = 2;
	uint32 total = 3;
}

message GetRowRequest {
	string table = 1;
	string pk = 2;
}

message GetRowResponse {
	repeated Column columns = 1;
	Row row = 2;
}

message WriteRowRequest {
	string table = 1;
	string pk = 2;
	map<string, string> fields = 3;
	bool update_only = 4;
	bool insert_only = 5;
}

message WriteRowResponse {}

message DeleteRowRequest {
	string table = 1;
	string pk = 2;
}

message DeleteRowResponse {}

message PersistRequest {}

message PersistResponse {}
